incubator/hnc/internal/reconcilers/anchor.go

```diff
@@ -161,69 +163,70 @@ func (r *AnchorReconciler) shouldDeleteSubns(log logr.Logger, inst *api.Subnames
 		return false
 	}
 
-	// If the anchor isn't in the "ok" state, don't delete the namespace. If it's "conflict," we
-	// *definitely* don't want to delete it; if it's missing (and possibly being created), then there
-	// isn't really a good option but we'll eventually put a condition on the namespace so it's not a
-	// big deal.
-	//
-	// TODO: much of the remaining portion of this function can probably be replaced by this switch
-	// statement. In v0.5, we're playing it safe so I won't modify the rest of the function, but in
-	// v0.6 we should restructure.
 	switch inst.Status.State {
 	case api.Missing:
+		// Either the namespace hasn't been created yet, or it's just been deleted. Either way, we don't
+		// need to delete it!
 		return false
 	case api.Conflict:
+		// The anchor is in a bad state; a namespace exists but isn't tied to this anchor. We certainly
+		// shouldn't delete the namespace.
 		return false
 	case api.Ok:
-		// keep going...
+		// The anchor and namespace are linked. We might still need to delete it.
 	default:
 		log.Error(errors.New("Bad anchor state"), "state", inst.Status.State)
 		// Stay on the safe side - don't delete
 		return false
 	}
 
-	cnm := inst.Name
-	pnm := inst.Namespace
-	cns := r.forest.Get(cnm)
-
-	// If the declared subnamespace is not created by this anchor, don't delete it.
-	if cns.Parent().Name() != pnm {
+	// If the namespace doesn't exist or is currently being deleted, we don't need to delete it again.
+	// Note that the state might be `Ok` (and not `Missing`) if the deleted namespace hasn't been
+	// reconciled yet, but there's no need to wait.
+	if nsInst.CreationTimestamp.IsZero() || !nsInst.DeletionTimestamp.IsZero() {
 		return false
 	}
```

Comment:
Seems right, since we now force setting parent in the forest based on `subof` annotation (source of truth). `cns.Parent().Name() != pnm` should be the same as `sOf != pnm` in the `updateState()` where `api.Conflict` state is set.